warehouse/query-core/src/test/java/datawave/query/predicate/KeyProjectionTest.java

```diff
@@ -0,0 +1,125 @@
+package datawave.query.predicate;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import org.apache.accumulo.core.data.Key;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class KeyProjectionTest {
+    
+    private List<Entry<Key,String>> fiData;
+    private List<Entry<Key,String>> eventData;
+    
+    @Before
+    public void setup() {
+        fiData = new ArrayList<>();
+        fiData.add(Maps.immutableEntry(new Key("20200314_1", "fi\0FIELD_A"), "data"));
+        fiData.add(Maps.immutableEntry(new Key("20200314_1", "fi\0FIELD_B"), "data"));
+        fiData.add(Maps.immutableEntry(new Key("20200314_1", "fi\0FIELD_C"), "data"));
+        fiData.add(Maps.immutableEntry(new Key("20200314_1", "fi\0FIELD_X"), "data"));
+        fiData.add(Maps.immutableEntry(new Key("20200314_1", "fi\0FIELD_Y"), "data"));
+        fiData.add(Maps.immutableEntry(new Key("20200314_1", "fi\0FIELD_Z"), "data"));
+        
+        eventData = new ArrayList<>();
+        eventData.add(Maps.immutableEntry(new Key("20200314_1", "datatype\0uid", "FIELD_A\0value_a"), "data"));
+        eventData.add(Maps.immutableEntry(new Key("20200314_1", "datatype\0uid", "FIELD_B\0value_b"), "data"));
+        eventData.add(Maps.immutableEntry(new Key("20200314_1", "datatype\0uid", "FIELD_C\0value_c"), "data"));
+        eventData.add(Maps.immutableEntry(new Key("20200314_1", "datatype\0uid", "FIELD_X\0value_x"), "data"));
+        eventData.add(Maps.immutableEntry(new Key("20200314_1", "datatype\0uid", "FIELD_Y\0value_y"), "data"));
+        eventData.add(Maps.immutableEntry(new Key("20200314_1", "datatype\0uid", "FIELD_Z\0value_z"), "data"));
+    }
+    
+    @Test(expected = RuntimeException.class)
+    public void testNoConfiguration() {
+        KeyProjection projection = new KeyProjection();
+        
+        Iterator<Entry<Key,String>> iter = fiData.iterator();
+        assertTrue(projection.apply(iter.next()));
+    }
+    
+    @Test(expected = RuntimeException.class)
+    public void testTooMuchConfiguration() {
+        KeyProjection projection = new KeyProjection();
+        projection.setIncludes(Sets.newHashSet("FIELD_A", "FIELD_B"));
+        projection.setExcludes(Sets.newHashSet("FIELD_X", "FIELD_Y"));
+        
+        Iterator<Entry<Key,String>> iter = fiData.iterator();
+        assertTrue(projection.apply(iter.next()));
+    }
+    
+    @Test(expected = RuntimeException.class)
+    public void testTooMuchOfTheSameConfiguration() {
+        KeyProjection projection = new KeyProjection();
+        projection.setExcludes(Sets.newHashSet("FIELD_X", "FIELD_Y"));
+        projection.setExcludes(Sets.newHashSet("FIELD_X", "FIELD_Y"));
```

Comment:
Looking at the code I understand why this fails, but why do we want it to fail?  I get that this is not a new concept - just trying to understand.  I'm guessing in practice it's never even set twice, but if it were, what would be the harm in just resetting to use whatever the last setter said to use (be it setInclude or setExclude)?

Reply:
I second this question.

Reply:
As far as I can tell, these projections were meant to be created once and this design decision was enforced by throwing an exception. Presumably this was to prevent state change in how the query handles includes and excludes?

Reply:
The includes/excludes should probably be immutable if that's the case and passed in as constructor arguments vs. using a setter.  Again, not really related to this PR, just an observation.  The approval still stands for me.

Reply:
Think I'd just like to see a new issue created for this. The scope of this ticket is really just updating the methods and adding unit tests for the existing behavior.

Reply:
Created https://github.com/NationalSecurityAgency/datawave/issues/1512 to update the contract for the Projection classes